towhee/serving/thread_model_serving.py

```diff
@@ -0,0 +1,231 @@
+        f = _TaskFuture(self._auto_task_id, len(data), weakref.ref(self.future_cache))
+
+        self.future_cache[f.task_id] = f
+
+        for idx, item in enumerate(data):
+            self._input_queue.put(_Task(f.task_id, idx, item))
+        return f
+
+    def __call__(self, data: Union['tensor', List['tensor']]) -> Union['tensor', List['tensor']]:
+        f = self._recv(data)
+        return f.result()
+
+    def get_batch_data(self):
+        wait_time = self._max_latency
+        batch_cache = []
+        while len(batch_cache) != self._batch_size and wait_time > 0:
+            time_start = time.time()
+            try:
+                if batch_cache:
+                    data = self._input_queue.get(timeout=wait_time)
+                else:
+                    data = self._input_queue.get()
+            except queue.Empty:
+                return batch_cache
+
+            if data is None:
+                # when return None, the serving is stopped and send a None to notify other workers.
+                self._input_queue.put(None)
+                return None
+
+            batch_cache.append(data)
+            wait_time = wait_time - (time.time() - time_start)
+
+        return batch_cache
+
+    def _loop(self, model):
+        while not self._need_stop:
+            batch_data = self.get_batch_data()
+            if batch_data is None:
+                break
+
+            batch_data_input = [item.content for item in batch_data]
+            batch_data_output = model(batch_data_input)
+            for i in range(len(batch_data_output)):
+                self.future_cache[batch_data[i].task_id].set_result(batch_data[i].sub_id, batch_data_output[i])
+
+    def start(self):
+        assert not self._need_stop, 'Serving only can starts once'
+
+        num_workers = len(self.models)
+        for replica_id in range(num_workers):
+            tworker = threading.Thread(target=self._loop, args=(self.models[replica_id],), daemon=True)
+            tworker.start()
+            self._tworkers.append(tworker)
+
+    def join(self):
+        for t in self._tworkers:
+            t.join()
+
+    def set_stop(self):
+        self._need_stop = True
+        self._input_queue.put(None)
+
+    def stop(self):
+        if self._need_stop:
+            return
+        self.set_stop()
+        self.join()
+
+    def proxy(self):
+        return ThreadServingProxy(self)
+
+    def __del__(self):
+        self.stop()
+
+
+class ThreadServingProxy:
+    def __init__(self, model_serving: ThreadModelServing):
+        self._model_serving = model_serving
+
+    @ property
+    def model(self):
+        return self._model_serving.model
+
+    def __call__(self, data: Union['tensor', List['tensor']]):
+        return self._model_serving(data)
+
+
+class _Task:
+    def __init__(self, task_id: int, sub_id: int, content: any):
+        self.task_id = task_id
+        self.sub_id = sub_id
+        self.content = content
+
+
+class _TaskFutureCache(dict):
```

Comment:
Is there a future plan for this `_TaskFutureCache`?

Reply:
The cache is just a dict, just give it another name for easy understanding.